```{r}
# libraries
# if (!requireNamespace("BiocManager", quietly = TRUE)){
#     install.packages("BiocManager")
#     BiocManager::install()
# }
# BiocManager::install("IsoformSwitchAnalyzeR")

library(IsoformSwitchAnalyzeR)
library(ggplot2)
library(dplyr)
library(tidyr)
```

```{r}
# format files
gtf_file <- "../results/isoquant/isoquant.transcript_models.gtf"
isoform_fasta <- "../results/transcriptome/transcriptome.fa"
output_path <- "../results/isoformswitchanalyzer"

new_col_names <- c("isoform_id","bc01_CTRL", "bc02_CTRL", "bc03_CTRL", "bc04_CTRL", "bc09_AD", "bc10_AD", "bc11_AD", "bc12_AD")
counts <- read.delim("../results/isoquant/isoquant.discovered_transcript_grouped_counts.tsv", sep = "\t")


names(counts) <- new_col_names
counts
```

```{r}
# abundance (transcripts per million)
# abundance <- read.delim("../results/isoquant/isoquant.discovered_transcript_grouped_tpm.tsv", sep = "\t")
# 
# names(abundance) <- new_col_names
# abundance
```

```{r}
# generate design matrix

sampleID <- colnames(counts[-1])
condition <- gsub('.*_', '', colnames(counts[-1])) %>% as.factor()
design <- data.frame(
    sampleID = sampleID,
    condition = condition
)

# add sex and age(?) and race(?) as covariates
design$sex <- factor(c('male','male','male','female','female','female','male','female'))

design
```

```{r}
# create switchAnalyzeRlist object
# calculated RPKM/FPKM if abundance not supplied

switch_analyzer_list <- importRdata(
    isoformCountMatrix   = counts,
    designMatrix         = design,
    isoformExonAnnoation = gtf_file,
    isoformNtFasta       = isoform_fasta,
    fixStringTieAnnotationProblem = TRUE,
    showProgress = TRUE
)
summary(switch_analyzer_list)
```

```{r}
head(switch_analyzer_list$isoformFeatures,5)
head(switch_analyzer_list$exons,5)
head(switch_analyzer_list$ntSequence,5)
```

```{r}
# filter:
# Multi-isoform genes
# Gene expression
# Isoform expression
# Isoform Fraction (isoform usage)
# Unwanted isoform classes
# Unwanted gene biotypes
# Genes without differential isoform usage

# default cutoffs:
# alpha=0.05 --> FDR
# dIFcutoff = 0.1 --> changes in absolute isoform usage / analogous to log2FC cut-off
# geneExpressionCutoff --> Default is 1 FPKM/TPM/RPKM
# isoformExpressionCutoff --> Default is 0 (which removes completely unused isoforms); also in RPKM/FPKM

switch_analyzer_list_filtered <- preFilter(
    switchAnalyzeRlist = switch_analyzer_list,
    geneExpressionCutoff = 1,
    isoformExpressionCutoff = 0,
    removeSingleIsoformGenes = TRUE
)
```

```{r}
# switch analysis with DEXseq
# default cutoffs:
# alpha=0.05 --> FDR
# dIFcutoff = 0.1 --> changes in absolute isoform usage / analogous to log2FC cut-off
# reduceToSwitchingGenes --> only uses genes w at least 1 sig differentially used isoform

switch_analyzer_list_filtered <- isoformSwitchTestDEXSeq(
    switchAnalyzeRlist = switch_analyzer_list_filtered,
    reduceToSwitchingGenes=TRUE
)
extractSwitchSummary(switch_analyzer_list_filtered)
```

```{r}
# analyze ORFs
switch_analyzer_list_filtered <- addORFfromGTF(
  switchAnalyzeRlist = switch_analyzer_list_filtered,
  pathToGTF = gtf_file,
  overwriteExistingORF=TRUE)

switch_analyzer_list_filtered <- analyzeNovelIsoformORF(
  switchAnalyzeRlist = switch_analyzer_list_filtered,
  analysisAllIsoformsWithoutORF = TRUE
) 
```

```{r}
# get amino acid seq
switch_analyzer_list_filtered <- extractSequence(
    switch_analyzer_list_filtered, 
    pathToOutput = output_path,
    writeToFile=TRUE
)
```
```{r}
switch_analyzer_list_filtered$orfAnalysis
```

```{r}
# downstream an analyses
switch_analyzer_list_filtered <- analyzeCPAT(
  switch_analyzer_list_filtered,
  file.path(output_path, "./")
)
```

----------------

### visualizations
```{r}
# gene-specific visualizations
switchPlot(
    switch_analyzer_list,
    gene='REC8',
    condition1 = 'CTRL',
    condition2 = 'AD',
    localTheme = theme_bw(base_size = 13) # making text sightly larger for vignette
)
```

```{r}
### genome-wide global usage of isoform switches
extractSwitchOverlap(
    switch_analyzer_list,
    filterForConsequences=TRUE,
    plotIsoforms = FALSE
)
```

## isoform switch consequence enrichment analysis
```{r}
# consequences of isoform switches by counting each type consequence separately
extractConsequenceSummary(
    switch_analyzer_list,
    consequencesToAnalyze='all',
    plotGenes = FALSE,           # enables analysis of genes (instead of isoforms)
    asFractionTotal = FALSE      # enables analysis of fraction of significant features
)
```

```{r}
# analysis of the predicted consequences of isoform switches
extractConsequenceEnrichment(
    switch_analyzer_list,
    consequencesToAnalyze='all',
    analysisOppositeConsequence = TRUE,
    localTheme = theme_bw(base_size = 14), # Increase font size in vignette
    returnResult = FALSE # if TRUE returns a data.frame with the summary statistics
)
```

```{r}
extractConsequenceEnrichmentComparison(
    switch_analyzer_list,
    consequencesToAnalyze=c('domains_identified','coding_potential'),
    analysisOppositeConsequence = TRUE,
    returnResult = FALSE # if TRUE returns a data.frame with the summary statistics
)
```

## splicing enrichment analyses
```{r}
extractSplicingEnrichment(
    switch_analyzer_list,
    returnResult = FALSE # if TRUE returns a data.frame with the summary statistics
)
```

```{r}
extractSplicingEnrichmentComparison(
    switch_analyzer_list,
    splicingToAnalyze = c('A3','A5','ATSS','ATTS'), # the splicing highlighted above
    returnResult = FALSE # if TRUE returns a data.frame with the results
)
```

```{r}
# volcano plot
ggplot(data=switch_analyzer_list$isoformFeatures, aes(x=dIF, y=-log10(isoform_switch_q_value))) +
     geom_point(
        aes( color=abs(dIF) > 0.1 & isoform_switch_q_value < 0.05 ), # default cutoff
        size=1
    ) +
    geom_hline(yintercept = -log10(0.05), linetype='dashed') + # default cutoff
    geom_vline(xintercept = c(-0.1, 0.1), linetype='dashed') + # default cutoff
    facet_wrap( ~ condition_2) +
    #facet_grid(condition_1 ~ condition_2) + # alternative to facet_wrap if you have overlapping conditions
    scale_color_manual('Signficant\nIsoform Switch', values = c('black','red')) +
    labs(x='dIF', y='-Log10 ( Isoform Switch Q Value )') +
    theme_bw()
```

```{r}
# Switch vs Gene changes
ggplot(data=switch_analyzer_list$isoformFeatures, aes(x=gene_log2_fold_change, y=dIF)) +
    geom_point(
        aes( color=abs(dIF) > 0.1 & isoform_switch_q_value < 0.05 ), # default cutoff
        size=1
    ) + 
    facet_wrap(~ condition_2) +
    #facet_grid(condition_1 ~ condition_2) + # alternative to facet_wrap if you have overlapping conditions
    geom_hline(yintercept = 0, linetype='dashed') +
    geom_vline(xintercept = 0, linetype='dashed') +
    scale_color_manual('Signficant\nIsoform Switch', values = c('black','red')) +
    labs(x='Gene log2 fold change', y='dIF') +
    theme_bw()
```

```{r}
# global consequence analysis
extractConsequenceSummary( switch_list )
extractConsequenceEnrichment( switch_list )
extractConsequenceGenomeWide( switch_list )

# global splicing analysis
extractSplicingSummary( switch_list )
extractSplicingEnrichment( switch_list )
extractSplicingGenomeWide( switch_list )
```

