---
title: "pacbio_ad_pilot_isoformswitchanalyzer_kallisto_counts"
author: Rachel Bozadjian
date: 8/6/25
output: 
  html_document:
    code_folding: show
    code_download: true
    df_print: paged
    theme: yeti
    highlight: tango
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: false
    number_sections: true
---

# Method

Transcript-level quantification was performed with lr-kallisto (v0.51.1), kallisto_optoff_k64 binary, using the GENCODE v48 human genome, transcriptome, and annotated GTF as references. Isoform abundance estimates (TxPM) and counts were imported into R where Differential Isoform Usage and splicing analyses were performed with IsoformSwitchAnalyzeR (v2.4.0).

The design matrix incorporated disease status (AD vs. CTRL) as the primary factor while adjusting for sex as a covariate. Filtering was performed with default parameters (alpha \< 0.05 and absolute isoform fraction difference (dIF) \> 0.1) to retain genes with expression ≥1 TPM and isoforms with non-zero expression.

Differential Isoform Usage was also assessed with significance thresholds of alpha \< 0.05 and absolute dIF \> 0.1. 26 total significant isoform switches were further characterized by integrating gene-level and isoform-level differential expression results from DESeq2 (v1.44.0). Functional consequences were predicted through four external computational tools: SignalP 6.0 for signal peptides, DeepTMHMM for protein topology, DeepLoc2.0 for subcellular localization, and Pfam for protein domains. We further evaluated the 26 significant isoform switches (isoform switch q value \< 0.05) for switching consequences, localization shifts, and alternative splicing following the IsoformSwitchAnalyzeR vignette with default parameters.

All analyses were conducted in R (v4.4.0) and tidyverse (v2.0.0) ecosystem packages. Plots were generated with IsoformSwitchAnalyzeR or ggplot2 (v3.5.2).

------------------------------------------------------------------------

```{r setup}
# libraries
library(ggrepel)
library(IsoformSwitchAnalyzeR)
library(tidyverse)

knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE, fig.width=12, fig.height=8, fig.align = "center")
```

```{r}
# print session information
sessionInfo()
```

# Import Kallisto Count Data

```{r}
# import counts
kallisto_counts_path <- "../kallisto_counts/"

kallisto_quant <- importIsoformExpression(
    parentDir = kallisto_counts_path,
    addIsofomIdAsColumn = TRUE
)

new_col_names <- c("isoform_id","bc01_CTRL", "bc02_CTRL", "bc03_CTRL", "bc04_CTRL", "bc09_AD", "bc10_AD", "bc11_AD", "bc12_AD")

names(kallisto_quant$abundance) <- new_col_names
names(kallisto_quant$counts) <- new_col_names
names(kallisto_quant$length) <- new_col_names

kallisto_quant$counts

# write.csv(kallisto_quant$counts, "kallisto_counts_all.csv", row.names = FALSE)
```

# Generate Design Matrix with Sex as Covariate

```{r}
# generate design matrix

sampleID <- colnames(kallisto_quant$abundance[-1])
condition <- gsub('.*_', '', colnames(kallisto_quant$abundance[-1])) %>% as.factor()
design <- data.frame(
    sampleID = sampleID,
    condition = condition
)

# add sex and ageand race as covariates
design$sex <- factor(c('male','male','male','female','female','female','male','female'))
#design$race <- factor(c('Black','Black', 'White', 'White', 'White','White','Black','Black'))

design
```

# Create Switch Analyzer Object

Used GENCODE v48 .gtf file and transcripts .fa file

```{r}
# create switchAnalyzeRlist object

gtf_file <- "../../refs/gencode.v48.annotation.gtf"
transcriptome_fasta <- "../../refs/gencode.v48.transcripts.fa.gz"

switch_analyzer_list <- importRdata(
    isoformCountMatrix   = kallisto_quant$counts,
    isoformRepExpression   = kallisto_quant$abundance,
    designMatrix         = design,
    isoformExonAnnoation = gtf_file,
    isoformNtFasta       = transcriptome_fasta,
    fixStringTieAnnotationProblem = TRUE,
    showProgress = TRUE
)
summary(switch_analyzer_list)
```

# Filtering

FDR \< 0.05

dIFcutoff = 0.1

geneExpressionCutoff =1 FPKM/TPM/RPKM

isoformExpressionCutoff = 0 RPKM/FPKM (removes completely unused isoforms)

```{r}
# filter:
# Multi-isoform genes
# Gene expression
# Isoform expression
# Isoform Fraction (isoform usage)
# Unwanted isoform classes
# Unwanted gene biotypes
# Genes without differential isoform usage

# default cutoffs:
# alpha=0.05 --> FDR
# dIFcutoff = 0.1 --> changes in absolute isoform usage / analogous to log2FC cut-off
# geneExpressionCutoff --> Default is 1 FPKM/TPM/RPKM
# isoformExpressionCutoff --> Default is 0 (which removes completely unused isoforms); also in RPKM/FPKM

switch_analyzer_list_filtered <- preFilter(
    switchAnalyzeRlist = switch_analyzer_list,
    geneExpressionCutoff = 1,
    isoformExpressionCutoff = 0,
    removeSingleIsoformGenes = TRUE
)
```

# Switch Analysis with DEXSeq

FDR \< 0.05

dIFcutoff = 0.1

reduceToSwitchingGenes = TRUE (only uses genes w at least 1 significant differentially used isoform)

```{r}
# switch analysis with DEXseq
# default cutoffs:
# alpha=0.05 --> FDR
# dIFcutoff = 0.1 --> changes in absolute isoform usage / analogous to log2FC cut-off
# reduceToSwitchingGenes --> only uses genes w at least 1 sig differentially used isoform

switch_analyzer_list_filtered <- isoformSwitchTestDEXSeq(
    switchAnalyzeRlist = switch_analyzer_list_filtered,
    reduceToSwitchingGenes=TRUE
)
extractSwitchSummary(switch_analyzer_list_filtered)
```

```{r}
# extract switch amino acid and nt sequences to files
switch_analyzer_list_filtered <- extractSequence(
    switch_analyzer_list_filtered, 
    pathToOutput = "results_kallisto_isoformanalyzer",
    writeToFile=TRUE
)
```

## Significant Isoform Switches

Arranged by isoform_switch_q_value

PTC = pre-mature termination codon

```{r}
# PTC means --> pre-mature termination codons

switch_analyzer_list_df <- switch_analyzer_list_filtered$isoformFeatures %>% 
  as.data.frame() %>% 
  arrange(isoform_switch_q_value)

switch_analyzer_list_df
#write.csv(switch_analyzer_list_df, "kallisto_isoform_switch_analysis.csv", row.names = FALSE)
```

## Add Differential Gene and Isoform Expression

Add padj values

```{r}
# add differential gene and isoform expression to switchanalyzer object

diff_gene_expr <- read.csv("../differential_expression/kallisto_DGE.csv")
diff_iso_expr <- read.csv("../differential_expression/kallisto_differential_isoform_expression.csv")

switch_analyzer_list_df <- switch_analyzer_list_df %>%
  left_join(
    diff_gene_expr %>% dplyr::select(gene_name, padj), 
    by = join_by(gene_id == gene_name)) %>%
  mutate(gene_q_value = padj) %>%
  subset(select = -padj) %>%
  left_join(
    diff_iso_expr %>% dplyr::select(isoform_id, padj), 
    by = "isoform_id") %>%
  mutate(iso_q_value = padj) %>%
  subset(select = -padj)

switch_analyzer_list_filtered$isoformFeatures$gene_q_value <- switch_analyzer_list_df$gene_q_value
switch_analyzer_list_filtered$isoformFeatures$iso_q_value <- switch_analyzer_list_df$iso_q_value

# save switch_analyzer_list_filtered as .RDS file for future use
#saveRDS(switch_analyzer_list_filtered, file = "switch_analyzer_list_filtered_sex.rds")
```

```{r}
# significant isoform switches (iso_switch_q_value < 0.05)
switch_analyzer_list_filtered <- readRDS("./switch_analyzer_list_filtered_sex.rds")
switch_analyzer_list_df <- switch_analyzer_list_filtered$isoformFeatures %>% as.data.frame()

significant_iso_switches <- switch_analyzer_list_df %>%
  dplyr::filter(isoform_switch_q_value < 0.05) %>%
  arrange(isoform_switch_q_value) %>%
  dplyr::select(isoform_id, gene_id, condition_1, condition_2,iso_biotype, IF1, IF2, dIF, isoform_switch_q_value, gene_switch_q_value, PTC)
```

# Volcano-like Plot of Significant Isoform Switches

reference is AD

isoform_switch_q_value \< 0.05

dIF \< -0.1 or \> 0.1

```{r}
# volcano-like plot of isoform switches AD vs CTRL
ggplot(data = switch_analyzer_list_df, aes(x = dIF, y = -log10(isoform_switch_q_value))) +
  geom_point(
    aes(color = abs(dIF) > 0.1 & isoform_switch_q_value < 0.05),
    size = 1
  ) +
  geom_hline(yintercept = -log10(0.05), linetype = 'dashed') +
  geom_vline(xintercept = c(-0.1, 0.1), linetype = 'dashed') +
  scale_color_manual('Significant\nIsoform Switch',
                     values = c('FALSE' = 'blue', 'TRUE' = 'orange')) +
  labs(x = 'dIF', y = '-Log10 (Isoform Switch Q Value)') +
  ggtitle("Significant Isoform Switches AD vs. CTRL (Q value < 0.05)") +
  ggrepel::geom_text_repel(
    aes(label = ifelse(abs(dIF) > 0.1 & isoform_switch_q_value < 0.05,
                       paste0(gene_id, " (", isoform_id, ")"),
                       '')),
    size = 2.5
  ) +
  theme_bw()
```

# External Analyses

signal peptides, protein topology, subcellular locations, and protein domains

```{r}
# SignalP - Prediction of Signal Peptides
# https://services.healthtech.dtu.dk/services/SignalP-6.0/
switch_analyzer_list_filtered <- analyzeSignalP(
    switchAnalyzeRlist       = switch_analyzer_list_filtered,
    pathToSignalPresultFile  = "./biolib_results/signalp/prediction_results.txt"
)

# analyzeDeepTMHMM : Prediction of protein topology — the proteins location compared to the cell membrane. Can be intracellular, transmembrane (TM) or extracellular.
# https://dtu.biolib.com/results/0797323f-c818-4343-88d7-43009062fc59/?token=ka2OLAD9XhWPRFIO
switch_analyzer_list_filtered <- analyzeDeepTMHMM(
    switchAnalyzeRlist   = switch_analyzer_list_filtered,
    pathToDeepTMHMMresultFile = "./biolib_results/deeptmhmm/TMRs.gff3",
    showProgress=TRUE
)

# analyzeDeepLoc2 - prediction of sub-cellular localization(s) of protein
# deeploc2.0
# https://services.healthtech.dtu.dk/services/DeepLoc-2.0/
switch_analyzer_list_filtered <- analyzeDeepLoc2(
    switchAnalyzeRlist = switch_analyzer_list_filtered,
    pathToDeepLoc2resultFile = "./biolib_results/deeploc2/results_6894F7A60000BCEA626E8931.csv",
    quiet = FALSE
)

# pfam
# see /bin/isoform_switch_analyzer/biolib_results/pfam_hmmscan/ directory for shell script
switch_analyzer_list_filtered <- analyzePFAM(
    switchAnalyzeRlist   = switch_analyzer_list_filtered,
    pathToPFAMresultFile = "./biolib_results/pfam_hmmscan/Pfam_result.txt",
    showProgress=TRUE
)
```

## Signal Peptide Analysis

```{r}
switch_analyzer_list_filtered$signalPeptideAnalysis %>% dplyr::left_join(switch_analyzer_list_filtered$isoformFeatures %>% dplyr::select(isoform_id, gene_name), by = join_by(isoform_id)) %>% dplyr::select(isoform_id, gene_name, everything())
```

## Topology Analysis

```{r}
switch_analyzer_list_filtered$topologyAnalysis %>% dplyr::left_join(switch_analyzer_list_filtered$isoformFeatures %>% dplyr::select(isoform_id, gene_name), by = join_by(isoform_id)) %>% dplyr::select(isoform_id, gene_name, everything())
```

## Subcellular Location Analysis

```{r}
switch_analyzer_list_filtered$subCellLocationAnalysis %>% dplyr::left_join(switch_analyzer_list_filtered$isoformFeatures %>% dplyr::select(isoform_id, gene_name), by = join_by(isoform_id)) %>% dplyr::select(isoform_id, gene_name, everything())
```

## Protein Domain Analysis

Sorted by E_value (significance)

```{r}
switch_analyzer_list_filtered$domainAnalysis %>% dplyr::left_join(switch_analyzer_list_filtered$isoformFeatures %>% dplyr::select(isoform_id, gene_name), by = join_by(isoform_id)) %>% dplyr::select(isoform_id, gene_name, everything()) %>% arrange(E_value)
```

# Isoform Switch Plots with External Analyses

```{r}
# isoform switch visualization

top_switches <- extractTopSwitches( 
    switch_analyzer_list_filtered, 
    filterForConsequences = FALSE, 
    n = NA,                  # n=NA: all features are returned
    extractGenes = FALSE,    # when FALSE isoforms are returned
    sortByQvals = TRUE,
)

plot_isoform_switch <- function(isoform_switch_list, gene) { switchPlot( isoform_switch_list, gene=gene, condition1 = 'AD', condition2 = 'CTRL', plotTopology = TRUE ) }

for (gene in top_switches$gene_id) { plot_isoform_switch(switch_analyzer_list_filtered, gene) }
```

# Alternative Splicing Analysis

Reference is AD

```{r}
# alternative splicing analysis
switch_analyzer_list_filtered <- analyzeAlternativeSplicing(
    switchAnalyzeRlist = switch_analyzer_list_filtered,
    quiet=TRUE
)
switch_analyzer_list_filtered$AlternativeSplicingAnalysis
```

```{r}
extractSplicingSummary( switch_analyzer_list_filtered )
extractSplicingEnrichment( switch_analyzer_list_filtered )
extractSplicingGenomeWide( switch_analyzer_list_filtered )
```

# Isoform Switch Consequence Analysis

```{r}
# switch consequences

consequencesOfInterest <- c(
        'intron_retention',
        'intron_structure',
        'ORF_genomic',
        'ORF_length',
        'NMD_status',
        'signal_peptide_identified',
        'last_exon',
        'exon_number',
        'tss',
        'tts',
        'isoform_seq_similarity',
        'isoform_length',
        "sub_cell_location",
        "sub_cell_shift_to_cell_membrane",
        "sub_cell_shift_to_cytoplasm",
        "sub_cell_shift_to_nucleus",
        "sub_cell_shift_to_Extracellular",
        "isoform_topology",
        "extracellular_region_count",
        "intracellular_region_count",
        "extracellular_region_length",
        "intracellular_region_length",
        '5_utr_seq_similarity',
        '5_utr_length',
        '3_utr_seq_similarity',
        '3_utr_length',
        'domains_identified',
        'domain_isotype',
        'domain_length',
        'genomic_domain_position'
    )

switch_analyzer_list_filtered <- analyzeSwitchConsequences(
    switch_analyzer_list_filtered,
    consequencesToAnalyze = consequencesOfInterest, 
    dIFcutoff = 0.1,
    showProgress=TRUE
)
```

## Top 10 Switches by Isoform Switch Q Value

```{r}
### Extract top switching genes (by q-value)
top_switches_q_value <- extractTopSwitches(
    switch_analyzer_list_filtered, 
    filterForConsequences = TRUE, 
    sortByQvals = TRUE
)

top_switches_q_value
```

## Top 10 Switches by dIF Value

```{r}
### Extract top switching genes (by dIF values)
top_switches_dif_value <- extractTopSwitches(
    switch_analyzer_list_filtered, 
    filterForConsequences = TRUE, 
    sortByQvals = FALSE
)
top_switches_dif_value
```

## Switch Consequence Plot AD vs CTRL

Reference is AD

```{r}
extractConsequenceEnrichment(
    switch_analyzer_list_filtered,
    consequencesToAnalyze="all",
    analysisOppositeConsequence = TRUE,
    localTheme = theme_bw(base_size = 14),
    returnResult = FALSE # if TRUE returns a data.frame with the summary statistics
)

```

## Plot of Subcellular Location Shifts

```{r}
extractSubCellShifts(
    switch_analyzer_list_filtered,
    plotGenes = FALSE,
    locationMinGenes = 1,
    returnResult = FALSE,
    localTheme = theme_bw()
)
```
